network/p2p/p2p.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-deadlock"
+
+	"github.com/libp2p/go-libp2p"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/muxer/yamux"
+	"github.com/libp2p/go-libp2p/p2p/transport/tcp"
+)
+
+// Service manages integration with libp2p
+type Service struct {
+	log       logging.Logger
+	host      host.Host
+	privKey   crypto.PrivKey
+	streams   *streamManager
+	pubsub    *pubsub.PubSub
+	pubsubCtx context.Context
+
+	topics   map[string]*pubsub.Topic
+	topicsMu deadlock.Mutex
+}
+
+// AlgorandWsProtocol defines a libp2p protocol name for algorand's websockets messages
+const AlgorandWsProtocol = "/algorand-ws/1.0.0"
+
+const dialTimeout = 30 * time.Second
+
+// MakeService creates a P2P service instance
+func MakeService(log logging.Logger, cfg config.Local, datadir string, pstore peerstore.Peerstore, wsStreamHandler StreamHandler) (*Service, error) {
+	// load stored peer ID, or make ephemeral peer ID
+	privKey, err := GetPrivKey(cfg, datadir)
+	if err != nil {
+		return nil, err
+	}
+
+	// muxer supports tweaking fields from yamux.Config
+	ymx := *yamux.DefaultTransport
+	// user-agent copied from wsNetwork.go
+	version := config.GetCurrentVersion()
+	ua := fmt.Sprintf("algod/%d.%d (%s; commit=%s; %d) %s(%s)", version.Major, version.Minor, version.Channel, version.CommitHash, version.BuildNumber, runtime.GOOS, runtime.GOARCH)
+
+	h, err := libp2p.New(
+		libp2p.Identity(privKey),
+		libp2p.UserAgent(ua),
+		libp2p.Transport(tcp.NewTCPTransport),
+		libp2p.Muxer("/yamux/1.0.0", &ymx),
+		libp2p.Peerstore(pstore),
+		libp2p.ListenAddrStrings("/ip4/0.0.0.0/tcp/0"), // TODO configuration
+		// libp2p.ConnectionGater(), // TODO
+	)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("P2P service started: peer ID %s addrs %s", h.ID(), h.Addrs())
+
+	sm := makeStreamManager(log, h, wsStreamHandler)
+	h.Network().Notify(sm)
+	h.SetStreamHandler(AlgorandWsProtocol, sm.streamHandler)
+
+	psCtx := context.TODO()
+	ps, err := makePubSub(psCtx, cfg, h)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		log:       log,
+		privKey:   privKey,
+		host:      h,
+		streams:   sm,
+		pubsub:    ps,
+		pubsubCtx: psCtx,
+		topics:    make(map[string]*pubsub.Topic),
+	}, nil
+}
+
+// Close shuts down the P2P service
+func (s *Service) Close() error {
+	return s.host.Close()
+}
+
+// Host returns the libp2p host
+func (s *Service) Host() host.Host {
+	return s.host
+}
+
+// DialPeers attempts to establish connections to the provided phonebook addresses
+func (s *Service) DialPeers(targetConnCount int) {
+	peerIDs := s.host.Peerstore().Peers()
+	for _, peerID := range peerIDs {
+		// if we are at our target count stop trying to connect
+		if len(s.host.Network().Conns()) >= targetConnCount {
```

Comment:
It might be possible to have connections exceed the target given multiple ways of connecting to others and the possibility of others connecting to us. I think it would be better to have this more robust condition than strict equality.